src/FuncTorrent/ControlThread.hs

```
@@ -0,0 +1,209 @@
+  -- Add delay here before polling PeerThreads again
+  checkAction
+
+ where
+   checkAction ct1 = do
+     putStrLn "Check action"
+     threadDelay $ 4*1000*1000
+     action <- readIORef $ view controlTAction ct1
+     case action of
+       FuncTorrent.ControlThread.Stop -> return ct1
+       _ -> mainLoop ct1
+
+doExit :: ControlThread -> IO ()
+doExit ct = do
+  putStrLn "Doing control-thread exit"
+  let peerTs = ct ^. peerThreads
+  -- let the peer threads stop themselves
+  mapM_ ((setPeerThreadAction FuncTorrent.PeerThreadData.Stop).fst) peerTs
+
+  -- Let the threads run for a while
+  -- We may add delay also if required
+  yield
+
+  -- remove all the threads which stopped successfully
+  ct1 <- clearFinishedThreads ct
+
+  -- If there are still threads waiting/blocked then either wait
+  -- if they are blocked due to disk write, then wait and retry
+  -- if thread not responding then kill the thread
+
+  unless (null (ct1 ^. peerThreads)) $ doExit ct1
+
+ where
+     clearFinishedThreads :: ControlThread -> IO ControlThread
+     clearFinishedThreads ct1 = do
+       remainingThreads <- filterM isRunning $ ct1 ^. peerThreads
+       return (ct1 & peerThreads .~ remainingThreads)
+      where
+          isRunning (_,tid) =
+              threadStatus tid >>= (\x -> return $ ThreadFinished /= x)
+
+getTrackerResponse :: ControlThread -> IO ControlThread
+getTrackerResponse ct = do
+  response <- tracker (ct^.metaInfo) "-HS0001-*-*-20150215"
+ 
+  -- TODO: Write to ~/.functorrent/caches
+  -- writeFile (name (info m) ++ ".cache") response
+ 
+  case decode response of
+    Right trackerInfo ->
+        case mkTrackerResponse trackerInfo of
+          Right trackerResp ->
+            let ct1 = trackerResponses %~ (trackerResp : ) $ ct
+                ct2 = peerList %~ ((peers trackerResp) ++) $ ct1
+            in return ct2
+          Left _ -> putStrLn "mkTracker error" >> return ct
+    Left _ -> putStrLn "tracker resp decode error" >> return ct
+
```

Both left cases seems wrong. Silent errors.